service/history/workflow/context.go

```diff
@@ -937,6 +937,8 @@ func (c *ContextImpl) UpdateRegistry(ctx context.Context, ms MutableState) updat
 				},
 			),
 		)
+	} else {
+		c.updateRegistry.UpdateFromStore()
```

Comment:
This is the key change: without this change, we encounter errors like the following:

1. We start and accept update 1 in cluster 1
2. We failover to cluster 2 and complete the update. This is fine so far.
3. We start update 2 in cluster 2, accept it, and fail back to cluster 1.
4. Now, we should be able to complete update 2 in cluster 1, but update 2 is not in the registry.

That's `TestAcceptedUpdateCanBeCompletedAfterFailoverAndFailback`. Another example is `TestConflictResolutionReappliesUpdatesSameIds`.

An alternative to this change might be to drop the registry at an appropriate time, in such a way that we get all the guarantees that we require. This PR does not explore that option; it is connected to work that @alexshtin is doing.

Reply:
Yes, I don't like the complexity (and `Futures` inconsistency) this PR adds. I am adding the ability to clear registry with proper notification of update result waiters. All this complexity will be removed and replaced with the call to `updateRegistry.Clear()`. For now, I am not against merging it.

Reply:
Yes, the main purpose of this PR is the added functional test coverage for updates and history replication. We need to merge it in order to ensure that future solutions don't break those tests.